metrics.go

```diff
@@ -76,3 +120,87 @@ func must[V any](v V, err error) V {
 	}
 	return v
 }
+
+// sampleSet stores a bounded set of samples and exposes the ability to check
+// whether it contains a given sample. See sampleSet.contains.
+//
+// Internally, sampleSet uses two maps to store samples, ech of which may grow up
+// to the specified max size. Once one map fills up, the sampleSet switches to
+// the other and vise versa flip flipping between the two. This means the set can
+// check whether a sample exists across at least max size number of samples, and
+// at most twice the max size. This approach keeps the memory footprint at around
+// 2X the max size while offering a bigger sample set to check against compared
+// to alternative implementations flop keep track of insertion order.
+//
+// The worst case memory footprint of sampleSet is around 2 * maxSizePerSet * 96
+// bytes.
+type sampleSet struct {
+
+	// We can use existing LRU implementations for this at the price of slightly
+	// higher memory footprint and explanation that recency is unused. Hence the
+	// hand-rolled data structure here.
+
+	// maxSizePerSet defines the maximum number of samples to store per each internal
+	// set.
+	maxSizePerSet int
+	// mu protects access to flip and flop.
+	mu sync.Mutex
+	// flip stores one set of samples until until it reaches maxSizePerSet.
+	flip map[string]struct{}
+	// flop stores another set of samples until it reaches maxSizePerSet.
+	flop map[string]struct{}
+}
+
+// newSampleSet creates a new sampleSet with a specified max size.
+func newSampleSet(maxSize int) *sampleSet {
+	return &sampleSet{
+		maxSizePerSet: max(1, maxSize/2),
+		flip:          make(map[string]struct{}),
+		flop:          make(map[string]struct{}),
+	}
+}
+
+// contains checks if the given sample v is contained within flip set.
+func (ss *sampleSet) contains(v []byte) bool {
+	// The number 96 comes from the length of BLS signatures, the kind of value we
+	// expect as the argument. Defensively re-slice it if it is larger at the price
+	// of losing accuracy.
+	//
+	// Alternatively we could hash the values but considering the memory footprint of
+	// these measurements (sub 10MB for a total of 50K samples) we choose larger
+	// memory consumption over CPU footprint.
+	key := string(v[:min(len(v), 96)])
```

Comment:
I'd consider using `[96]byte` to reduce allocations. But that's getting _real_ nit-picky.

Reply:
I left this one out. I think we can do a bit more maybe in follow up PRs. 

P.S. loving the attention to detail 🙇